metrics-exporter/src/bin/io_engine/collector/pool_stat.rs

```diff
@@ -0,0 +1,198 @@
+        let pool_bytes_written_opts = Opts::new("bytes_written", "Total bytes written on the pool")
+            .subsystem("diskpool")
+            .variable_labels(vec!["node".to_string(), "name".to_string()]);
+        let pool_num_write_ops_opts =
+            Opts::new("num_write_ops", "Number of write operations on the pool")
+                .subsystem("diskpool")
+                .variable_labels(vec!["node".to_string(), "name".to_string()]);
+        let pool_read_latency_opts =
+            Opts::new("read_latency", "Total read latency on the pool in usec")
+                .subsystem("diskpool")
+                .variable_labels(vec!["node".to_string(), "name".to_string()]);
+        let pool_write_latency_opts =
+            Opts::new("write_latency", "Total write latency on the pool in usec")
+                .subsystem("diskpool")
+                .variable_labels(vec!["node".to_string(), "name".to_string()]);
+        let mut descs = Vec::new();
+
+        let pool_bytes_read = GaugeVec::new(pool_bytes_read_opts, &["node", "name"])
+            .expect("Unable to create gauge metric type for pool_total_size");
+        let pool_num_read_ops = GaugeVec::new(pool_num_read_ops_opts, &["node", "name"])
+            .expect("Unable to create gauge metric type for pool_used_size");
+        let pool_bytes_written = GaugeVec::new(pool_bytes_written_opts, &["node", "name"])
+            .expect("Unable to create gauge metric type for pool_committed_size");
+        let pool_num_write_ops = GaugeVec::new(pool_num_write_ops_opts, &["node", "name"])
+            .expect("Unable to create gauge metric type for pool_total_size");
+        let pool_read_latency = GaugeVec::new(pool_read_latency_opts, &["node", "name"])
+            .expect("Unable to create gauge metric type for pool_used_size");
+        let pool_write_latency = GaugeVec::new(pool_write_latency_opts, &["node", "name"])
+            .expect("Unable to create gauge metric type for pool_committed_size");
+        // Descriptors for the custom metrics
+        descs.extend(pool_bytes_read.desc().into_iter().cloned());
+        descs.extend(pool_num_read_ops.desc().into_iter().cloned());
+        descs.extend(pool_bytes_written.desc().into_iter().cloned());
+        descs.extend(pool_num_write_ops.desc().into_iter().cloned());
+        descs.extend(pool_read_latency.desc().into_iter().cloned());
+        descs.extend(pool_write_latency.desc().into_iter().cloned());
+
+        Self {
+            pool_bytes_read,
+            pool_num_read_ops,
+            pool_bytes_written,
+            pool_num_write_ops,
+            pool_read_latency,
+            pool_write_latency,
+            descs,
+        }
+    }
+}
+
+impl Collector for PoolIoStatsCollector {
+    fn desc(&self) -> Vec<&prometheus::core::Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<prometheus::proto::MetricFamily> {
+        let mut c = match Cache::get_cache().lock() {
```

Comment:
Why don't we store the ref counted cache in the collector?

Reply:
All Collectors are initialized when /metrics is called by client. So we access Cache using OnceCell struct when /metrics is called.